Implement unify_from_right

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22308 More named inference rules for pointwise unary ops
* #22268 Named inference rule for more pointwise ops.
* #22267 Enable named inference for some unary pointwise ops
* **#22223 Implement unify_from_right**
* #22151 Named inference rule for `abs`.

This is the named inference rule for broadcasting binary pointwise
operations, like add. For example, in `tensor + other`,
`unify_from_right(tensor.names(), other.names())` gives the names of the
output tensor of the operation.

A few notes:
1. Doesn't support tagged names yet. It's not too hard to support but
the error messages must be nice.
2. Error messages can be improved to suggest workarounds. If the tensors
have the same names but just in a different order (alignment), we could
suggest the user to use `tensor + other.align_as(tensor)`.
3. This is O(n^2) where n is the number of dimensions but for N < 64
creating a std::set or std::unordered_set isn't worth it.

Test Plan:
- `build/bin/NamedTensor_test` [namedtensor ci]

gh-metadata: pytorch pytorch 22223 gh/zou3519/62/head

Differential Revision: [D16030551](https://our.internmc.facebook.com/intern/diff/D16030551)

zou3519 committed:
Implement unify_from_right

This is the named inference rule for broadcasting binary pointwise
operations, like add. A few notes:

1. Doesn't support tagged names yet. It's not too hard to support but
the error messages must be nice.
2. Error messages can be cleaned up. Future work.
3. This is O(n^2) where n is the number of dimensions but for N < 64
creating a std::set or std::unordered_set isn't worth it.

Test Plan:
- `build/bin/NamedTensor_test` [namedtensor ci]

zou3519 committed:
Update on "Implement unify_from_right"

Implement unify_from_right

This is the named inference rule for broadcasting binary pointwise
operations, like add. For example, in `tensor + other`,
`unify_from_right(tensor.names(), other.names())` gives the names of the
output tensor of the operation.

A few notes:
1. Doesn't support tagged names yet. It's not too hard to support but
the error messages must be nice.
2. Error messages can be cleaned up. Future work.
3. This is O(n^2) where n is the number of dimensions but for N < 64
creating a std::set or std::unordered_set isn't worth it.

Test Plan:
- `build/bin/NamedTensor_test` [namedtensor ci]

gh-metadata: pytorch pytorch 22223 gh/zou3519/62/head

zou3519 committed:
Update on "Implement unify_from_right"

Implement unify_from_right

This is the named inference rule for broadcasting binary pointwise
operations, like add. For example, in `tensor + other`,
`unify_from_right(tensor.names(), other.names())` gives the names of the
output tensor of the operation.

A few notes:
1. Doesn't support tagged names yet. It's not too hard to support but
the error messages must be nice.
2. Error messages can be improved to suggest workarounds. If the tensors
have the same names but just in a different order (alignment), we could
suggest the user to use `tensor + other.align_as(tensor)`.
3. This is O(n^2) where n is the number of dimensions but for N < 64
creating a std::set or std::unordered_set isn't worth it.

Test Plan:
- `build/bin/NamedTensor_test` [namedtensor ci]

gh-metadata: pytorch pytorch 22223 gh/zou3519/62/head

zou3519 committed:
Update on "Implement unify_from_right"

Implement unify_from_right

This is the named inference rule for broadcasting binary pointwise
operations, like add. For example, in `tensor + other`,
`unify_from_right(tensor.names(), other.names())` gives the names of the
output tensor of the operation.

A few notes:
1. Doesn't support tagged names yet. It's not too hard to support but
the error messages must be nice.
2. Error messages can be improved to suggest workarounds. If the tensors
have the same names but just in a different order (alignment), we could
suggest the user to use `tensor + other.align_as(tensor)`.
3. This is O(n^2) where n is the number of dimensions but for N < 64
creating a std::set or std::unordered_set isn't worth it.

Test Plan:
- `build/bin/NamedTensor_test` [namedtensor ci]

gh-metadata: pytorch pytorch 22223 gh/zou3519/62/head

nairbv commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -51,6 +51,95 @@ int64_t dimname_to_position(const Tensor& tensor, Dimname dim) {
+}
+
+// Assumption: A DimnameList can have no duplicate full names with
+// the exception of wildcards
```
I think it'd be helpful to have comments explaining what this function is for, maybe explaining the behavior too or linking to any docs on it

zou3519 replied:
Sure, I'll toss some examples in.

zou3519 replied:
Added docs in NamedTensorUtils.h

nairbv commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -51,6 +51,95 @@ int64_t dimname_to_position(const Tensor& tensor, Dimname dim) {
+  }
+  auto it = std::find_if(other_names.begin(), other_names.end(),
+      [&](const Dimname& candidate) { return name.can_refer_to(candidate); });
+  // XXX: Can improve message by checking if names are alignable and suggesting workarounds
```
why 'XXX' instead of 'TODO'?

zou3519 replied:
It's easier to type. I don't know what the difference is between "XXX" and "TODO", but I'll change it to TODO so that it actually means something.

nairbv commented on aten/src/ATen/Dimname.h:
```diff
@@ -20,6 +20,10 @@ struct CAFFE2_API Dimname {
 
   bool can_refer_to(const Dimname& other) const;
 
+  bool is_normal() const { return type_ == NameType::NORMAL; }
```
I don't have a better suggestion right now but we may eventually want to come up with a more explicit name for 'normal'. I feel like the word 'normal' doesn't convey any information without knowing that the other options are wildcard and tagged.

zou3519 replied:
I'd like to rename it to "simple" sometime in the future

zou3519 committed:
Update on "Implement unify_from_right"

Implement unify_from_right

This is the named inference rule for broadcasting binary pointwise
operations, like add. For example, in `tensor + other`,
`unify_from_right(tensor.names(), other.names())` gives the names of the
output tensor of the operation.

A few notes:
1. Doesn't support tagged names yet. It's not too hard to support but
the error messages must be nice.
2. Error messages can be improved to suggest workarounds. If the tensors
have the same names but just in a different order (alignment), we could
suggest the user to use `tensor + other.align_as(tensor)`.
3. This is O(n^2) where n is the number of dimensions but for N < 64
creating a std::set or std::unordered_set isn't worth it.

Test Plan:
- `build/bin/NamedTensor_test` [namedtensor ci]

gh-metadata: pytorch pytorch 22223 gh/zou3519/62/head

zou3519 committed:
Update on "Implement unify_from_right"

Implement unify_from_right

This is the named inference rule for broadcasting binary pointwise
operations, like add. For example, in `tensor + other`,
`unify_from_right(tensor.names(), other.names())` gives the names of the
output tensor of the operation.

A few notes:
1. Doesn't support tagged names yet. It's not too hard to support but
the error messages must be nice.
2. Error messages can be improved to suggest workarounds. If the tensors
have the same names but just in a different order (alignment), we could
suggest the user to use `tensor + other.align_as(tensor)`.
3. This is O(n^2) where n is the number of dimensions but for N < 64
creating a std::set or std::unordered_set isn't worth it.

Test Plan:
- `build/bin/NamedTensor_test` [namedtensor ci]

gh-metadata: pytorch pytorch 22223 gh/zou3519/62/head
